torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,138 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// kRpcRetryDuration is the initial duration between consecutive RPC retries.
+const auto kRpcRetryDuration = std::chrono::milliseconds(1000);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+const float kBackoff = 1.5;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  if (rpcAgentRunning_.load()) {
+    shutdown();
+  }
+}
+
+void RpcAgent::shutdown() {
+  std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  lock.unlock();
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  auto future = std::make_shared<FutureMessage>();
+  steady_clock_time_point newTime = computeNewRpcRetryTime(/* retryCount */ 0);
+  auto fm = send(to, std::move(message));
+
+  fm->addCallback([this, future, newTime](
+                      const rpc::Message& message,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
+    } else if (!future->completed()) {
+      // First try succeeded and we can mark the original future as completed.
+      future->markCompleted(message);
+      {
+        std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+        rpcRetryMap_.erase(newTime);
+      }
+    }
+  });
+
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
```

Comment:
Instead of adding the RpcRetryInfo to the map, we can move this logic to the callback. Basically, if in the future callback we receive an error we add the entry to the map.

The same thing can be done in retryExpiredRpcs where we pick the first RPC to retry, remove it from the map, retry it and in the callback decide to add it back to the map if we haven't exceeded maxRetries. This way we automatically clean up the map and also no longer need `shouldRetryRpc` and `retryFuture_`